src/bin/ambit/cmd.rs

```diff
@@ -52,24 +53,122 @@ fn is_symlinked(link_name: &Path, target: &Path) -> bool {
         .unwrap_or(false)
 }
 
+fn get_paths_from_spec(spec: &Spec, start_path: PathBuf) -> AmbitResult<Vec<PathBuf>> {
+    let mut paths: Vec<PathBuf> = Vec::new();
+    let contains_pattern_char = |s: &str| -> bool { s.contains('*') || s.contains('?') };
+    for entry in spec.into_iter() {
+        if contains_pattern_char(&entry) {
+            fn get_matches(
+                ancestor_path: &Path,
+                expected_path_kind: &AmbitPathKind,
+                does_match: impl Fn(&str) -> bool,
+            ) -> AmbitResult<Vec<PathBuf>> {
+                let mut matches = Vec::new();
+                // Iterate through all entries in ancestor_path.
+                for path in fs::read_dir(ancestor_path)? {
+                    let path = path?.path();
+                    // Validify the current path.
+                    if let Some(file_name) = path.file_name() {
+                        if match expected_path_kind {
+                            AmbitPathKind::File => path.is_file(),
+                            AmbitPathKind::Directory => path.is_dir(),
+                        } && does_match(&file_name.to_string_lossy().to_string())
+                        {
+                            matches.push(path);
+                        }
+                    }
+                }
+                Ok(matches)
+            }
+            // The only valid path at the start is the starting path.
+            let mut valid_paths: Vec<PathBuf> = vec![start_path.clone()];
```

Comment:
You should probably explain that `valid_paths` gets replaced with `new_valid_paths` at every depth. (The algorithm in general should be explained.)